auth-service/src/main/kotlin/uk/gov/communities/delta/auth/utils/RegistrationService.kt

```diff
@@ -0,0 +1,155 @@
+package uk.gov.communities.delta.auth.utils
+
+import OrganisationService
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.config.DeltaConfig
+
+class RegistrationService(
+    private val deltaConfig: DeltaConfig,
+    private val organisationService: OrganisationService,
+    private val emailService: EmailService
+) {
+    class RegistrationResult {
+        enum class Status {
+            CREATED, ALREADY_EXISTS, OTHER_FAILURE
+        }
+
+        val status: Status
+        val email: String
+        val token: String?
+
+        constructor(
+            status: Status,
+            email: String
+        ) {
+            this.status = status
+            this.email = email
+            token = null
+        }
+
+        constructor(
+            status: Status,
+            email: String,
+            token: String
+        ) {
+            this.status = status
+            this.email = email
+            this.token = token // TODO - are these still needed with tokens stored in db (see below)
+        }
+    }
```

Comment:
You can do something like

```kotlin
sealed class RegistrationResult
class ResultCreated(val email: String, val token: String): RegistrationResult
class ResultAlreadyExists(val email: String): RegistrationResult
class ResultFailure(val exception: Exception?): RegistrationResult
```

then

```kotlin
when (result) {
  is ResultCreated -> { doSomethingWith(result.token) }
  ...
}
```

`sealed` means the compiler knows there aren't any other instances of the class so it can check the `when` is exhaustive